src/descriptionBuilder.js

```diff
@@ -0,0 +1,105 @@
+const { getEmojiUrl, imageToBase64 } = require("./utils");
+
+const DescriptionBuilder = function(desc, remove_emojis = false) {
+  let x = 25; // start x position of description
+  let maxX = 80; // description will be stripped if x exceeds 90 (= about 55 letters without emojis)
+  const y = 70; // vertical positioning of description line
+  let parentType = null; // text, image
+
+  // stores description parts
+  const descriptionRenderParts = [];
+
+  function addText(text, force = false) {
+    if (parentType === 'image') {
+      x += 2; // extra spacing between text and image
+    }
+    if (x+text.length > maxX && !force) {
+      text = `${text.slice(0, maxX-x)}..`;
+    }
+    descriptionRenderParts.push('<text class="description" x="' + x + '" y="' + y + '">' + text + '</text>');
+    x += text.length;
+    parentType = 'text';
+  }
+
+  function _addImage(imageUrl, width, height) {
+    descriptionRenderParts.push('<image xmlns:xlink="http://www.w3.org/1999/xlink" width="' + width + 'px" height="' + height + 'px" x="' + x + '" y="' + (parseInt(y)-12) + '" xlink:href="' + imageUrl + '" />');
+    x += width;
+    maxX += Math.ceil(width * 0.8); // approximate value to increase maxX as more characters can be displayed when image was added
+    parentType = 'image';
+  }
+
+  function addImage(imageUri, width, height) {
+    return new Promise(async (resolve) => {
+      if (imageUri === null) {
+        resolve();
+        return;
+      }
+      if (!imageUri.startsWith('http')) {
+        _addImage(imageUri, width, height);
```

Comment:
> I think instead of mutating the descriptionParts array in here and pushing directly to it, we should make the _addText & _addImage function just return......

Following up on this comment let push to descriptionParts here instead of doing that on each function method, its more clear this way.